### PR TITLE
Road crossings

### PR DESCRIPTION
**Pull request by Skleni**
Adds support for the different road crossing types (T, X, symmetric Y, asymmetric Y)

grafik

Thanks to @bumchen for the geometric support :)

### CODE REVIEW

File: src/OpenSage.Game/Graphics/Rendering/RenderPipeline.cs [hunk 1]
```diff
@@ -224,8 +224,20 @@ public void Execute(RenderContext context)
...
+            commandList.PushDebugGroup("Terrain");
+            RenderedObjectsOpaque += DoRenderPass(context, commandList, _renderList.Terrain, standardPassCameraFrustum, cloudResourceSet);
+            commandList.PopDebugGroup();
+
+            commandList.PushDebugGroup("Road");
+            RenderedObjectsOpaque += DoRenderPass(context, commandList, _renderList.Road, standardPassCameraFrustum, cloudResourceSet);
+            commandList.PopDebugGroup();
+
+            commandList.PushDebugGroup("Water");
+            DoRenderPass(context, commandList, _renderList.Water, standardPassCameraFrustum, cloudResourceSet);
```
Comment (feliwir) [inline src/OpenSage.Game/Graphics/Rendering/RenderPipeline.cs]: Why no `RenderedObjectsOpaque +=` here?

File: src/OpenSage.Game/Terrain/Roads/CrossingRoadSegment.cs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+            var rightSideDirection = Vector3.Normalize(rightDirection - upDirection);
+            var leftSideDirection = Vector3.Normalize(-rightDirection - upDirection);
+
+            var targetSize = GetSize(RoadTextureType.SymmetricYCrossing, template);
+
+            var lengthToTop = 0.269f * targetSize.Height;
+            var lengthToBottom = targetSize.Height - lengthToTop;
+
+            var halfRoadWidth = template.RoadWidth * template.RoadWidthInTexture / 2f;
+            var lengthToSide = 0.547f * template.RoadWidth;
```
Comment (feliwir) [inline src/OpenSage.Game/Terrain/Roads/CrossingRoadSegment.cs]: Do all these magic constants have some mathematical background or are they just defined like this by the engine?

File: src/OpenSage.Game/Terrain/Roads/RoadSegmentMesher.cs [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        }
+
+        protected override Vector3 ToTopBorder(float relativeProgress)
+        {
+            return HalfHeight * DirectionNormalNoZ;
+        }
+
+        protected override float TopUOffset(float relativeProgress)
+        {
+            return 0;
```
Comment (feliwir) [inline src/OpenSage.Game/Terrain/Roads/RoadSegmentMesher.cs]: Is this a TODO?

File: src/OpenSage.Game/Terrain/Roads/TextureAtlas.cs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+        // The length of the incoming roads in T and X crossings in texture space.
+        private const float IncomingRoadStubLength = 0.004f;
+
+        // The length from the center to the endpoint of the incoming roads in T and X crossings in texture space. 
+        private const float LengthToIncomingRoad = UnscaledRoadWidth / 2 + IncomingRoadStubLength;
+
+        // There are 3x3 tiles in a road texture, so a half tile is 1/6 = 0.1666...,
+        // but the calculations are based on the rounded number.
+        private const float HalfTileSize = 0.166f;
```
Comment (feliwir) [inline src/OpenSage.Game/Terrain/Roads/TextureAtlas.cs]: Shouldn't the rounded number be 0.167f?

File: src/OpenSage.Game/Graphics/Rendering/RenderPipeline.cs (hunk 1 above)
Comment (Skleni) [reply src/OpenSage.Game/Graphics/Rendering/RenderPipeline.cs]: I'm not sure, I copied this from #353...maybe water doesn't count as opaque?

File: src/OpenSage.Game/Terrain/Roads/CrossingRoadSegment.cs (hunk 2 above)
Comment (Skleni) [reply src/OpenSage.Game/Terrain/Roads/CrossingRoadSegment.cs]: No idea to be honest...we tried to deduce and name as much as possible, but some magic numbers are left.

File: src/OpenSage.Game/Terrain/Roads/RoadSegmentMesher.cs (hunk 3 above)
Comment (Skleni) [reply src/OpenSage.Game/Terrain/Roads/RoadSegmentMesher.cs]: No, this is intended :)

File: src/OpenSage.Game/Terrain/Roads/TextureAtlas.cs (hunk 4 above)
Comment (Skleni) [reply src/OpenSage.Game/Terrain/Roads/TextureAtlas.cs]: Yeah, it should :D It isn't though...for whatever reason. I used the unrounded 1/6 for a long time and never know why the coordinates were a little off.

**Pull request merged**